### DISCUSSION TITLE
How to add custom validators that adhere to spectree's validation?

### DISCUSSION THREAD

**Discussion by guilhermeschmitd-hotmart** [0b01001001/spectree#361] [Q&A] [answered]
I have model that requires further validation according to business rules, such as the following example. When this validation fails, I'd like my API to return error messages in the same way `spectree` does for missing or incorrect fields.

```python
class Item(BaseModel):
    item_id: int
    name: Annotated[str, Field(min_length=3)]
    amount: Annotated[int, Field(gt=0)]
    price: Annotated[float, Field(gt=0)]

class ShoppingCart(BaseModel):
    items: Annotated[Sequence[Item], Field(min_items=1)]
    
    @validator('items')
    @classmethod
    def validate_items(cls, value):
        if len({v['item_id'] for v in value}) < len(value):
                raise Exception('A cart cannot have duplicate items. Use amounts instead.')  # What should I do here?
        
        return value
```

Can this be achieved or should I use a separate validation step within the endpoint logic?

**Comment by kemingy** [comment, marked as answer]
You can raise `ValueError` here. It will be converted to a ValidationError in pydantic. Then it will be processed by spectree like the usual validation errors.

**Reply by guilhermeschmitd-hotmart** [reply]
Works like a charm, I was so close 😅 
Thank you!
